crates/synthesis/src/discrete_basis/math.rs

```diff
@@ -0,0 +1,284 @@
+    let angle = ((trace - 1.) / 2.).acos();
+    let lhs = (angle / 2.).sin();
+
+    // we use a bisect search to solve the angle equation (Eq. 10 in the paper)
+    let min_angle = 0.;
+    let max_angle = 2. * (1. / 2_f64.powf(0.25)).asin();
+
+    let f = move |phi: f64| -> f64 {
+        let sin_sq = (phi / 2.).sin().powi(2);
+        2. * sin_sq * (1. - sin_sq.powi(2)).sqrt() - lhs
+    };
+
+    bisect(f, min_angle, max_angle, 1e-20, 70).expect("Who coded this cannot math")
+}
+
+/// Add coeff * V to the out matrix, where V is the skew-symmetric representation of the
+/// cross product with the vector v.
+fn add_cross_prod(coeff: f64, v: &Matrix3x1<f64>, out: &mut Matrix3<f64>) {
+    out[(0, 1)] -= coeff * v[2];
+    out[(0, 2)] += coeff * v[1];
+    out[(1, 0)] += coeff * v[2];
+    out[(1, 2)] -= coeff * v[0];
+    out[(2, 0)] -= coeff * v[1];
+    out[(2, 1)] += coeff * v[0];
+}
+
+/// Add coeff * v.dot(v.T) to the output matrix.
+fn add_outer_prod(coeff: f64, v: &Matrix3x1<f64>, out: &mut Matrix3<f64>) {
+    out[(0, 0)] += coeff * v[0] * v[0];
+    out[(0, 1)] += coeff * v[0] * v[1];
+    out[(0, 2)] += coeff * v[0] * v[2];
+    out[(1, 0)] += coeff * v[1] * v[0];
+    out[(1, 1)] += coeff * v[1] * v[1];
+    out[(1, 2)] += coeff * v[1] * v[2];
+    out[(2, 0)] += coeff * v[2] * v[0];
+    out[(2, 1)] += coeff * v[2] * v[1];
+    out[(2, 2)] += coeff * v[2] * v[2];
+}
+
+/// Compute the group commutator UVU^{-1}V^{-1}. This code leverages the fact that
+/// the matrices as SO(3) and U^{-1} = U^T.
+fn group_commutator(left: &Matrix3<f64>, right: &Matrix3<f64>) -> Matrix3<f64> {
+    left * right * left.transpose() * right.transpose()
+}
+
+/// Compute the rotation axis of a SO(3) matrix.
+fn rotation_axis_from_so3(matrix: &Matrix3<f64>, tol: f64) -> Matrix3x1<f64> {
+    let trace = matrix.trace();
+
+    // Try to obtain the matrix via Rodrigues formula. If the matrix is diagonal (or the Rodrigues
+    // approach fails, likely because the matrix was diagonal up to numerical error), determine
+    // the axis from the diagonal.
+    if trace >= 3. - tol {
+        // matrix is the identity, ie. no rotation
+        return Matrix3x1::identity();
+    }
+
+    if trace >= tol - 1. {
+        // there's a skew symmetric part which we can use to get the rotation axis
+        let theta = ((trace - 1.) / 2.).acos();
+        let skew = (matrix - matrix.transpose()) / 2.;
+
+        if theta.sin() > tol {
+            let coeff = 1. / 2. / theta.sin();
+            let axis = Matrix3x1::new(
+                coeff * (skew[(2, 1)] - skew[(1, 2)]),
+                coeff * (skew[(0, 2)] - skew[(2, 0)]),
+                coeff * (skew[(1, 0)] - skew[(0, 1)]),
+            );
+
+            // this might fail due to numerical error, in that case go to diagonal case
+            if !axis.iter().any(|el| el.is_nan()) {
+                return axis / vector_norm(&axis);
+            }
+        }
+    }
```

Comment:
Given our discussions of the Python version of this function, I would expect this block to be something like the following. Note that we should use `matrix` and not `skew`. I think that the check `theta.sin() > tol` is redundant. I also don't quite see how `!axis.iter().any(|el| el.is_nan())` can fail and why you need to manually normalize axis by its norm (it should already be correct based on the computation).


```suggestion
    if trace >= tol - 1. {
        // there's a skew symmetric part which we can use to get the rotation axis
        let theta = ((trace - 1.) / 2.).acos();
        let coeff = 1. / 2. / theta.sin();
        let axis = Matrix3x1::new(
            coeff * (matrix[(2, 1)] - matrix[(1, 2)]),
            coeff * (matrix[(0, 2)] - matrix[(2, 0)]),
            coeff * (matrix[(1, 0)] - matrix[(0, 1)]),
        );

        return axis;
    }
```

Reply:
Yeah I think this can only break if someone passes a _very_ small `tol`, in which case `theta.sin()` could evaluate to 0 and things explode. If we can guarantee that `tol` is large enough we can remove these checks

Reply:
I am almost certain this should follow automatically, the condition `trace >= tol - 1.` means that $\theta$ is $tol$ away from both $0$ and $\pi$ (using $\sin(\theta) \approx \theta$ for $\theta \approx 0$).

Reply:
I agree with your math, but I'm pretty sure I put that check there for a reason -- I can play with it after fixing the other things and get back to you 🙂

Reply:
You're right, it doesn't fail even if I remove the tolerance check. I'm still mildly nervous about removing this, since I'm mainly testing on RZ gates, so I'd prefer keeping it as safe guard if you don't mind.

Reply:
I don't mind, even if I don't think you need this.